packages-exp/auth-exp/src/core/strategies/popup.ts

```diff
@@ -0,0 +1,249 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as externs from '@firebase/auth-types-exp';
+import { FirebaseError } from '@firebase/util';
+
+import { Auth } from '../../model/auth';
+import {
+  AuthEvent,
+  AuthEventConsumer,
+  AuthEventType,
+  EventManager,
+  PopupRedirectResolver
+} from '../../model/popup_redirect';
+import { User, UserCredential } from '../../model/user';
+import { AUTH_ERROR_FACTORY, AuthErrorCode } from '../errors';
+import { Delay } from '../util/delay';
+import { _generateEventId } from '../util/event_id';
+import { AuthPopup } from '../util/popup';
+import { _link, _reauth, _signIn, IdpTask, IdpTaskParams } from './idp';
+
+// The event timeout is the same on mobile and desktop, no need for Delay.
+export const _AUTH_EVENT_TIMEOUT = 2020;
+export const _POLL_WINDOW_CLOSE_TIMEOUT = new Delay(2000, 10000);
+
+interface PendingPromise {
+  resolve: (cred: UserCredential) => void;
+  reject: (error: Error) => void;
+}
+
+export async function signInWithPopup(
+  authExtern: externs.Auth,
+  provider: externs.AuthProvider,
+  resolverExtern: externs.PopupRedirectResolver
+): Promise<UserCredential> {
+  const auth = authExtern as Auth;
+  const resolver = resolverExtern as PopupRedirectResolver;
+
+  const action = new PopupAction(
+    auth,
+    AuthEventType.SIGN_IN_VIA_POPUP,
+    _signIn,
+    provider,
+    resolver
+  );
+  const cred = await action.execute();
+
+  await auth.updateCurrentUser(cred.user);
```

Comment:
If you wait for my refactor, _signIn calls signInWithCredential and does the updateCurrentUser for you, then all three of these look the same

Reply:
Done